### ISSUE TITLE
`useFormContext` should be available through children

### ISSUE THREAD

**Issue by wladiston** [SeasonedSoftware/remix-forms#37]
Since we are using react-hook-form. It would be nice to have access to stuff like `isSubmitting` from the form children without having to pass it down the three.

Right now, the form is controlling the label of the button to indicate the form is submiting but if you are working with UI libs like Chakra, you might want to use somethnig like ```<Button isLoading />```

**Comment by danielweinmann** [comment]
@wladiston @bvangraafeiland, I just added `isLoading` to the [button on our Chakra UI example](https://github.com/SeasonedSoftware/remix-forms-chakra-ui-example/blob/main/app/ui/button.tsx) to demonstrate how we can accomplish that without `useFormContext`.

However, that solution will not work if you're using the `fetcher` prop. Because of that, I created #122 to create an even more sustainable solution.

**Comment by danielweinmann** [comment]
Regarding `useFormContext`, all [methods](https://react-hook-form.com/api/useformcontext) returned by it are already passed to the children of `Form`, including `formState`, which has the `isSubmitting` value.

That said, I wouldn't use RHF's state to check for the pending state in Remix or React Router 6.4. I would [useTransition](https://remix.run/docs/en/v1/api/remix#usetransition), [useNavigation](https://reactrouter.com/en/main/hooks/use-navigation), or [useFetcher](https://reactrouter.com/en/main/hooks/use-fetcher).

**Comment by bvangraafeiland** [comment]
> Regarding `useFormContext`, all [methods](https://react-hook-form.com/api/useformcontext) returned by it are already passed to the children of `Form`, including `formState`, which has the `isSubmitting` value.
> 
> That said, I wouldn't use RHF's state to check for the pending state in Remix or React Router 6.4. I would [useTransition](https://remix.run/docs/en/v1/api/remix#usetransition), [useNavigation](https://reactrouter.com/en/main/hooks/use-navigation), or [useFetcher](https://reactrouter.com/en/main/hooks/use-fetcher).

I'm not seeing those props? This is what I see when logging the props of Input
<img width="350" alt="Screenshot 2022-12-07 at 14 11 35" src="https://user-images.githubusercontent.com/3051636/206188024-bf941618-457e-4945-976e-7753e9043be8.png">

And this for Button:
<img width="340" alt="Screenshot 2022-12-07 at 14 12 27" src="https://user-images.githubusercontent.com/3051636/206188165-8ad1001c-7154-4ff0-b9ac-3489b07df9b4.png">

**Comment by danielweinmann** [comment]
@bvangraafeiland, here's how you access the [UseFormReturn](https://react-hook-form.com/ts#UseFormReturn):

```tsx
<Form schema={schema}>
    {({ Field, Errors, Button, register, formState /* and everything in UseFormReturn, all type-safe to the schema */ }) => (<>/* Your markup */</>)}
  </Form>
```

**Comment by bvangraafeiland** [comment]
Yes, but that's meant to be used on a per-form basis, right? I think I'm not explaining this properly. What I'm looking for is a way to have the submit button:
- Have the `disabled` prop set based on a combination of Remix's `transition` and `formState.isValid`
- While submitting, the button's content should be a loading spinner - there is the `pendingButtonLabel` prop, but it's of type `string`. I guess this could be changed to `ReactNode` but in the end there is more flexibility if the custom component could access state through a hook.
- This behavior should be enabled by default on all forms that are created through my `Form` component based on `createForm`

Currently I have this behavior for a single form like this:
```
{({ Field, Button, Errors, formState }) => (
  <>
    <Field name="email" />
    <Field name="password" type="password" />
    <Errors />
    <Button disabled={!!transition.submission || (!formState.isValid && formState.submitCount > 0)}>
      Log in
    </Button>
  </>
)}
```
But that would need to be copied for every form. Having `useFormState` and `useField` available in custom components would make this, but also things like styling for errors, a lot easier. Currently I'm (ab)using the aria attributes being put on the input components for errors with tailwind using `aria-invalid:border-red-600`.

**Comment by danielweinmann** [comment]
Yeah, that's the use case I had in mind when creating #122. Your idea of adding would also work, and I'd love to explore both :) if you create issues for them we'll explore that path as well.

**Comment by danielweinmann** [comment]
It took me some time, but I finally understood this issue 😅

If we always render a [FormProvider](https://react-hook-form.com/api/formprovider) inside our forms, it will enable the users to [useFormContext](https://react-hook-form.com/api/useformcontext) anywhere inside any form. The solution is simple and looks harmless. I'll reopen the issue until we implement it.

**Comment by bvangraafeiland** [comment]
Yeah that would definitely solve my use case 👍

**Comment by danielweinmann** [comment]
> Yeah that would definitely solve my use case 👍 If it helps, I could try opening a PR for it?

Definitely! We'd really appreciate a PR, @bvangraafeiland 💪🏼

**Comment by danielweinmann** [comment]
Released on [v1.3.0](https://github.com/SeasonedSoftware/remix-forms/releases/tag/v1.3.0). Thanks for the PR, @bvangraafeiland 💪🏼 🎉
